pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -320,12 +321,45 @@ class Utils {
                     tagsAction.addTag(tagName, tagValue)
                     currentNode.save()
                 }
+            }
+        }
+    }
+
+    static void markStartAndEndNodesInStageAsNotExecuted(String stageName, FlowExecution execution = null) {
```

Comment:
So, if I understand correctly, we're hunting for the start/end nodes for ALL stages and and parallels labelled with stageName?  There may be a nuance here I'm missing but here's a simpler way I'd propose rather than doing the multi-pass search:

```java
public void markStageAndBranchNotExecuted(FlowExecution exec, String labelName) {
        DepthFirstScanner scan = new DepthFirstScanner();
        Predicate MATCH = new Predicate<FlowNode>() {
            @Override
            public boolean apply(@Nullable FlowNode x) {
                if (x instanceof BlockStartNode) {
                    LabelAction label = x.getAction(LabelAction.class);
                    // Works because ParallelLabelAction extends LabelAction implements ThreadNameAction
                    // And stages will have a StageAction too
                    if (label instanceof ThreadNameAction || x.getAction(StageAction.class) != null) {
                        return labelName.equals(label.getDisplayName());
                    }
                } else if (x instanceof BlockEndNode) {
                    FlowNode f = ((StepEndNode) x).getStartNode();
                    return this.apply(f);
                }
                return false;
            }
        };
        scan.setup(exec.getCurrentHeads());
        scan.filter(MATCH).forEachRemaining(x -> ((FlowNode)x).addAction(new NotExecutedNodeAction()));
    }
```
Yes, it will do an exhaustive scan (unfortunately) but I think it's somewhat unavoidable with this logic. 

Note: I haven't tested the above, so there may be some edge case I'm missing

Reply:
Note that only works with block-scoped stages -- IIRC we don't use any other kind with Declarative though, right?

Reply:
Correct, no block-scoped. I'll try the `GraphListener` approach first, because that just does seem smarter if I can pull it off.

Reply:
If you do GraphListener you should be able to use the predicate -- just replace the string equality with checking a `HashSet<String>` for the label -- note this relies on not being able to have collisions between executed and not executed names for any stage/parallel branch.